smithy-typescript-ssdk-libs/server-common/src/index.ts

```diff
@@ -0,0 +1,45 @@
+/*
+ *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ *  A copy of the License is located at
+ *
+ *   http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+export * as httpbinding from "./httpbinding";
+export * from "./errors";
+
+import { HttpRequest, HttpResponse } from "@aws-sdk/protocol-http";
+import { SmithyException } from "@aws-sdk/smithy-client";
+import { SerdeContext } from "@aws-sdk/types";
+
+export type Operation<I, O> = (input: I, request: HttpRequest) => Promise<O>;
+
+export type OperationInput<T> = T extends Operation<infer I, any> ? I : never;
+export type OperationOutput<T> = T extends Operation<any, infer O> ? O : never;
+
+export interface OperationSerializer<T, K extends keyof T, E extends SmithyException> {
+    serialize(input: OperationOutput<T[K]>, ctx: Omit<SerdeContext, 'endpoint'>): Promise<HttpResponse>
+    deserialize(input: HttpRequest, ctx: SerdeContext): Promise<OperationInput<T[K]>>
+    isOperationError(error: any): error is E
+    serializeError(error: E, ctx: Omit<SerdeContext, 'endpoint'>): Promise<HttpResponse>
+}
+
+export interface ServiceHandler {
```

Comment:
Just curious: why not use the [RequestHandler interface in v3](https://github.com/AllanZhengYP/aws-sdk-js-v3/blob/b344ed8edb8146bbca3d7481168d889eb69fe1b5/packages/types/src/transfer.ts#L9)? Because `metadata` or `destroy` is not needed?

Reply:
I'm not sure - @adamthom-amzn might know since he wrote this bit. From looking at it it does look like it could tie us into having client metadata / options.

Reply:
omg typescript lets you have default values for type parameters that's amazing

Reply:
Didn't know that it existed, to be honest. Not sure we want to tie the protocol to the handler implementation, though I haven't thought about it that much.